File: core/src/main/java/y2k/joyreactor/presenters/PostListPresenter.java
package y2k.joyreactor.presenters;

import rx.Observable;
import y2k.joyreactor.*;
import y2k.joyreactor.common.Messages;
import y2k.joyreactor.platform.Navigation;
import y2k.joyreactor.services.repository.Repository;
import y2k.joyreactor.services.TagService;
import y2k.joyreactor.services.synchronizers.PostListSynchronizer;

import java.util.List;

/**
 * Created by y2k on 9/26/15.
 */
public class PostListPresenter extends Presenter {

    private View view;
    private TagService serviceFactory;
    private TagService service;

    public PostListPresenter(View view) {
        this(view, new TagService(new Repository<>(Post.class),
                new PostListSynchronizer.Factory(new Repository<>(Post.class), new Repository<>(TagPost.class))));
    }

    @Deprecated
    PostListPresenter(View view,
                      Repository<Post> repository,
                      PostListSynchronizer.Factory synchronizerFactory) {
        this(view, new TagService(repository, synchronizerFactory));
    }

    PostListPresenter(View view, TagService serviceFactory) {
        this.view = view;
        this.serviceFactory = serviceFactory;

        getMessages().add(this::currentTagChanged, Messages.TagSelected.class);
        currentTagChanged(new Messages.TagSelected(Tag.makeFeatured()));
    }

    private void currentTagChanged(Messages.TagSelected m) {
        service = serviceFactory.make(m.tag);

        view.setBusy(true);
        getFromRepository().subscribe(posts -> view.reloadPosts(posts, null));
        service.preloadNewPosts()
                .subscribe(unsafeUpdate -> {
                    view.setHasNewPosts(unsafeUpdate);
                    view.setBusy(false);
                    if (!unsafeUpdate) applyNew();
                });
    }

    public void applyNew() {
        service.applyNew()
                .subscribe(posts -> {
                    view.setHasNewPosts(false);
                    view.reloadPosts(posts, service.getDivider());
                });
    }

    public void loadMore() {
        view.setBusy(true);
        service.loadNextPage()
                .subscribe(posts -> {
                    view.reloadPosts(posts, service.getDivider());
                    view.setBusy(false);
                });
    }

    public void reloadFirstPage() {
        view.setBusy(true);
        service.reloadFirstPage()
                .subscribe(posts -> {
                    view.reloadPosts(posts, posts.size());
                    view.setBusy(false);
                });
    }

    private Observable<List<Post>> getFromRepository() {
        return service.queryAsync();
    }

    public void postClicked(Post post) {
        Navigation.getInstance().openPost(post.serverId);
    }

    public void playClicked(Post post) {
        if (post.image.isAnimated()) Navigation.getInstance().openVideo(post.serverId);
        else Navigation.getInstance().openImageView(post);
    }

    public interface View {

        void setBusy(boolean isBusy);

        void reloadPosts(List<Post> posts, Integer divider);

        void setHasNewPosts(boolean hasNewPosts);
    }
}

File: core/src/main/kotlin/y2k/joyreactor/presenters/PostListPresenter.kt
package y2k.joyreactor.presenters

import rx.Observable
import y2k.joyreactor.Post
import y2k.joyreactor.Tag
import y2k.joyreactor.common.subscribeOnMain
import y2k.joyreactor.platform.Navigation
import y2k.joyreactor.services.BroadcastService
import y2k.joyreactor.services.LifeCycleService
import y2k.joyreactor.services.TagService

/**
 * Created by y2k on 9/26/15.
 */
class PostListPresenter(
    private val view: PostListPresenter.View,
    private val service: TagService,
    private val lifeCycleService: LifeCycleService) {

    init {
        lifeCycleService.add(BroadcastService.TagSelected::class) { currentTagChanged(it.tag) }
        currentTagChanged(Tag.makeFeatured())
    }

    private fun currentTagChanged(newTag: Tag) {
        service.setTag(newTag)

        view.setBusy(true)
        getFromRepository().subscribeOnMain { view.reloadPosts(it, null) }

        service
            .preloadNewPosts()
            .subscribeOnMain { unsafeUpdate ->
                view.setHasNewPosts(unsafeUpdate)
                view.setBusy(false)
                if ((!unsafeUpdate)) applyNew()
            }
    }

    fun applyNew() {
        service.
            applyNew()
            .subscribeOnMain { posts ->
                view.setHasNewPosts(false)
                view.reloadPosts(posts, service.divider)
            }
    }

    fun loadMore() {
        view.setBusy(true)
        service
            .loadNextPage()
            .subscribeOnMain { posts ->
                view.reloadPosts(posts, service.divider)
                view.setBusy(false)
            }
    }

    fun reloadFirstPage() {
        view.setBusy(true)
        service
            .reloadFirstPage()
            .subscribeOnMain { posts ->
                view.reloadPosts(posts, posts.size)
                view.setBusy(false)
            }
    }

    private fun getFromRepository(): Observable<List<Post>> {
        return service.queryAsync()
    }

    fun postClicked(post: Post) {
        Navigation.instance.openPost(post.serverId!!)
    }

    fun playClicked(post: Post) {
        if (post.image!!.isAnimated) Navigation.instance.openVideo(post.serverId!!)
        else Navigation.instance.openImageView(post)
    }

    interface View {

        fun setBusy(isBusy: Boolean)

        fun reloadPosts(posts: List<Post>, divider: Int?)

        fun setHasNewPosts(hasNewPosts: Boolean)
    }
}